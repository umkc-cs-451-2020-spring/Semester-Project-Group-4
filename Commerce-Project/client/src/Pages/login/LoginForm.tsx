import React from 'react'
import { Link as RouterLink, LinkProps as RouterLinkProps } from 'react-router-dom';
import { TextField, Button, FormGroup, Grid, Typography, Paper } from '@material-ui/core'
import { useFormik } from 'formik';
import { makeStyles, createStyles } from '@material-ui/styles';
import apis from '../../api';
import { Redirect } from 'react-router-dom';
import { useLocalStorageSetState } from '../../utils/local-storage';

const useStyles = makeStyles(() =>
  createStyles({
    paper: {
      padding: '1rem',
      height: '100%',
      margin: '5rem 1rem',
    },
    textFieldMargin: {
      margin: '.5rem'
    },
    title: {
      marginBottom: '1rem'
    },
    buttonMargin: {
      marginTop: '2.5rem'
    },
    button: {
      marginTop: '1rem'
    },
    gridMargin: {
      margin: 'unset',
      width: '100%'
    },
    font: {
      fontSize: 0
    }
  }),
);

export const validate = (values: any) => {
  const errors = {} as any;
  if (!values.username) {
    errors.username = 'Required';
  } else if (values.username.length > 20) {
    errors.username = 'Invalid username';
  }
  if (!values.password) {
    errors.password = 'Required';
  } else if (!/^(?=.*[a-z])(?=.*[A-Z])(?=.*\d).{8,}$/i.test(values.password)) {
    errors.password = 'Invalid password.';
  }
  return errors;
};

const LoginForm = () => {
  const { textFieldMargin, buttonMargin, gridMargin, button, title, font, paper } = useStyles();
  const [next, setNext] = React.useState(false)
  const [name, setName] = useLocalStorageSetState('', 'username')

  const registerLink = React.useMemo(
    () =>
      React.forwardRef<any, Omit<RouterLinkProps, 'to'>>((itemProps, ref) => (
        <RouterLink to='/register' ref={ref} {...itemProps} />
      )),
    []
  );

  const formik = useFormik({
    initialValues: {
      username: '',
      password: ''
    },
    validate,
    onSubmit: (values: any) => {
      apis.getUser(values).then(() => {
        setName(formik.values.username);
        formik.setSubmitting(false);
        setNext(true)
        window.location.reload()
      })
    }
  });

  return (
    <form onSubmit={formik.handleSubmit}>
      <Paper elevation={0} className={paper}>
        <Grid container spacing={8} className={gridMargin}>
          <Grid item xs={4} />
          <Grid item xs={4}>
            <FormGroup>
              <Typography className={title} align='center' variant="h5">Log in to Online Banking</Typography>
              <TextField
                className={textFieldMargin}
                label="Username"
                id="username"
                name="username"
                type="text"
                onChange={formik.handleChange}
                value={formik.values.username}
                helperText={formik.errors.username && formik.touched.username ? formik.errors.username : null}
                error={formik.touched.username && formik.errors.username ? true : false}
                onBlur={formik.handleBlur}
              />
              <TextField
                className={textFieldMargin}
                label="Password"
                id="password"
                name="password"
                type="password"
                onChange={formik.handleChange}
                value={formik.values.password}
                helperText={formik.errors.password && formik.touched.password ? formik.errors.password : null}
                error={formik.touched.password && formik.errors.password ? true : false}
                onBlur={formik.handleBlur}
              />
              <Button className={buttonMargin} type="button" variant="contained" color="primary" onClick={formik.submitForm}>Login</Button>
              <Button className={button} type="button" variant="contained" color="inherit" component={registerLink}>Register</Button>
            </FormGroup>
          </Grid>
          <Typography className={font}>{name}</Typography>
        </Grid>
      </Paper>
      {next ? <Redirect to="/dashboard" /> : false}
    </form >
  )
}

export default LoginForm
