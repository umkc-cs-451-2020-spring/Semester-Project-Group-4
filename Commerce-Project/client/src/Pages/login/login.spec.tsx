import React from 'react';
import { shallow } from 'enzyme';
import LoginForm, { validate } from './LoginForm';

describe('Login Form', () => {

  it('Render the Login Form', () => {
    const component = shallow(<LoginForm />);

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });


  it('Validate should throw value required', () => {

    const form = {
      username: '',
      firstName: '',
      lastName: '',
      email: '',
      phone: '',
      password: '',
      confirmPassword: ''
    };

    const expectedValue = {
      "username": "Required",
      "password": "Required"
    }

    expect(validate(form)).toEqual(expectedValue);
  });

  it('Validate should throw length errors', () => {

    const invalidLongName = 'BobBobBobBobBobBobBobBobBobBobBobBobBobBobBobBobBobBobBob'

    const form = {
      username: invalidLongName,
      password: 'password!'
    };

    const expectedValue = {
      "password": "Invalid password.",
      "username": "Invalid username"
    }

    expect(validate(form)).toEqual(expectedValue);
  });
});

