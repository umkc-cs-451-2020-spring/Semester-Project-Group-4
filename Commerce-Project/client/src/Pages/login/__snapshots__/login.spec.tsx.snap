// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form Render the Login Form 1`] = `
<form
  onSubmit={[Function]}
>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-paper-1"
    elevation={0}
  >
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-gridMargin-6"
      container={true}
      spacing={8}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={4}
      />
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={4}
      >
        <WithStyles(ForwardRef(FormGroup))>
          <WithStyles(ForwardRef(Typography))
            align="center"
            className="makeStyles-title-3"
            variant="h5"
          >
            Log in to Online Banking
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(TextField))
            className="makeStyles-textFieldMargin-2"
            error={false}
            helperText={null}
            id="username"
            label="Username"
            name="username"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
          <WithStyles(ForwardRef(TextField))
            className="makeStyles-textFieldMargin-2"
            error={false}
            helperText={null}
            id="password"
            label="Password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value=""
          />
          <WithStyles(ForwardRef(Button))
            className="makeStyles-buttonMargin-4"
            color="primary"
            onClick={[Function]}
            type="button"
            variant="contained"
          >
            Login
          </WithStyles(ForwardRef(Button))>
          <WithStyles(ForwardRef(Button))
            className="makeStyles-button-5"
            color="inherit"
            component={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              }
            }
            type="button"
            variant="contained"
          >
            Register
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(FormGroup))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-font-7"
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Paper))>
</form>
`;
